src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php

```diff
@@ -125,7 +128,7 @@ public function preSubmit(FormEvent $event)
                 if (!$form->has($name)) {
                     $form->add($name, $this->type, array_replace(array(
                         'property_path' => '['.$name.']',
-                    ), $this->options));
+                    ), $this->getOptionsForEntry($value)));
```

Comment:
The view data is not the problem :) Think about when we need to call that closure:

* During `setData()`: Here we want to create the fields for a collection in model format
* During `submit()`: Here we want to create the fields for an array in view format

It is technically impossible (it seems to me) to solve this problem with one closure that accepts only either the model format or the view format. Dealing with both formats in the closure, on the other hand, is not really usable.